# Redes Neuronales

# Hoja de trabajo # 2

### Rodrigo Samayoa

### 17332

Imports necesarios

In [32]:
import mnist_reader
import numpy as np
from scipy import optimize as op
from functools import reduce
from funcionesRedes import *

Lectura de CSV's

In [33]:
X_train, y_train = mnist_reader.load_mnist('', kind='train')
X_test, y_test = mnist_reader.load_mnist('', kind='t10k')

Normalizacion de data y obtención de dimensiones

In [34]:
X = X_train/1000.0
m,n = X.shape
y = y_train.reshape(m,1)
Y = (y == np.array(range(10))).astype(int)

In [35]:
NET = np.array([
        n,
        130,
        10
    ])
theta_shapes = np.hstack((
    NET[1:].reshape(len(NET)-1, 1),
    (NET[:-1]+1).reshape(len(NET)-1,1)
))


flat_thetas = flatten_list_of_arrays([
    np.random.rand(*theta_shape)*0.01
    for theta_shape in theta_shapes
])

# Desarrollo de Red Neuronal

No correr esta parte del codigo ya que tarda entre 3 a 5 horas.

-----------------------------------------------------------------------------

In [ ]:
print("\n---------------- OPTIMIZING ----------------\n")
result = op.minimize(
    fun = cost_function,
    x0 = flat_thetas,
    args = (theta_shapes, X, Y),
    method = 'L-BFGS-B',
    jac = prediction_calculation,
    options = {'disp': True, 'maxiter': 900}
)
print("\n---------------- OPTIMIZED ----------------\n")

Se guardan las thetas en un txt

In [ ]:
np.savetxt('thetas_optimizadas.txt', result.x,delimiter=',') 
#(NO correr: dará error porque la variable result (que se define arriba) 
#no existe ya que se tiene que correr lo de arriba primero)

-----------------------------------------------------------------------------

Se cargan los thetas del txt

In [36]:
thetaResult = np.loadtxt('thetas_optimizadas.txt')
thetaResult.size

103360

# Prueba

In [37]:
X_test = X_test/1000
mT,nT = X_test.shape
y_test = y_test.reshape(mT,1)

In [38]:
result_flat_thetas = thetaResult

In [39]:
super_thetas_result = inflate_matrixes(result_flat_thetas, theta_shapes)
a_result = feed_forward(super_thetas_result, X_test)

In [40]:
maximos = np.argmax(a_result[-1], axis = 1)
accuracy = 0
for i in range(len(a_result[-1])):
    if(maximos[i] == y_test[i][0]):
        accuracy += 1
success = accuracy*100/len(a_result[-1])
failure = 100.0 - success
print("éxito:", success, "%")
print("fallo:", failure, "%")

éxito: 88.22 %
fallo: 11.780000000000001 %


# Analisis

Como se puede observar el porcentaje de exito del modelo fue de 88.22%, por lo que se puede decir que el modelo es un buen modelo ya que entreno de una manera correcta. El modelo consiste de una capa oculta con 130 neuronas, solo se utilizo una capa oculpa ya que se nos dijo que el modelo no tenia que ser tan complejo. 